### PR TITLE
Improve code efficiency to handle large amount of data + add plotting utils

### PR DESCRIPTION
**Pull request by JoachimPiret**
Dataset.py : improve code efficiency of two functions of class AggregatedDataset() :  init() and  mapSplitsToDataset()
transformtomonthly.py : allows to record dataframe in parquet format in addition to csv format
dataloader.py : add possibility to divide between test and train absed on subregion (c-region) as well as the possibility to have randomness and different sampling from sampling to sampling in set_train_test_split().  assign_train_test_indices(self,train_indices, test_indices, test_size) is defined to update dataloader with the values of the selected test/train divisions after 10 sampling based on subregion.

### CODE REVIEW

File: massbalancemachine/data_processing/Dataset.py [hunk 1]
```diff
@@ -391,44 +394,48 @@ def __init__(
...
             ]
         )
-        self.uniqueID = np.unique(self.ID)
-        self.maxConcatNb = max(
-            [len(np.argwhere(self.ID == id)[:, 0]) for id in self.uniqueID]
-        )
+        # self.uniqueID = np.unique(self.ID)
+        # self.maxConcatNb = max(
+        #     [len(np.argwhere(self.ID == id)[:, 0]) for id in self.uniqueID]
+        # ) #take a lot of time with a lot of data, replaced with the 2 following lines
```
Comment (albangossard) [inline massbalancemachine/data_processing/Dataset.py]: That's cleaner, thanks!
You can remove the commented lines
```suggestion
```

File: massbalancemachine/data_processing/Dataset.py [hunk 2]
```diff
@@ -391,44 +394,48 @@ def __init__(
...
-
+    
         Returns:
             list[tuple[np.ndarray, np.ndarray]]: List with the same number of tuples
                 as the input. Each tuple contains numpy arrays which provide the
                 corresponding indices the cross validation should use according to
                 the input splits variable.
         """
+        # Precompute the mapping of unique IDs to indices
+        uniqueID_to_indices = {uid: np.where(self.uniqueID == uid)[0] for uid in np.unique(self.uniqueID)}
```
Comment (albangossard) [inline massbalancemachine/data_processing/Dataset.py]: Elements of `self.uniqueID` are already unique
```suggestion
        uniqueID_to_indices = {uid: np.where(self.uniqueID == uid)[0] for uid in self.uniqueID}
```

Comment (albangossard) [reply massbalancemachine/data_processing/Dataset.py]: I don't think that what you are actually computing corresponds to what you want: here `uniqueID_to_indices` is equivalent to the identity (up to permutations because of `np.unique(self.uniqueID)`). What you want instead is probably something like
```python
uniqueID_to_indices = {uid: np.where(self.ID == uid)[0] for uid in self.uniqueID}
```

File: massbalancemachine/dataloader/DataLoader.py [hunk 3]
```diff
@@ -64,15 +64,16 @@ def __init__(
         self.test_indices = None
         self.meta_data_columns = meta_data_columns or cfg.metaData
 
+    
     def set_train_test_split(
-        self, *, test_size: float = None, type_fold: str = "group-meas-id"
+        self, *, test_size: float = None, type_fold: str = "group-meas-id", randomness : bool = False
```
Comment (albangossard) [inline massbalancemachine/dataloader/DataLoader.py]: `randomness` is a poor name since when we set it to `False`, there is still a random shuffling. `random_state` or `random_seed` would be more appropriate

File: massbalancemachine/dataloader/DataLoader.py [hunk 4]
```diff
@@ -108,8 +116,20 @@ def set_train_test_split(
         # Make it iterators and set as an attribute of the class
         self.train_indices = train_indices
         self.test_indices = test_indices
-
         return iter(self.train_indices), iter(self.test_indices)
+    
+    def assign_train_test_indices(self,train_indices, test_indices, test_size):
```
Comment (albangossard) [inline massbalancemachine/dataloader/DataLoader.py]: What is the purpose of this function? I don't see it being called in the code and its description doesn't match with what it does. Is it intended? Or is it a WIP?

**Review by albangossard — CHANGES_REQUESTED** [review]
Hi @JoachimPiret, thanks for opening this PR! The changes in `mapSplitsToDataset` will probably improve the performance on large datasets, that's great!
There are a few things that need to be fixed.
Note that I didn't run the code since the execution of the notebooks is currently failing in the CI. It looks like there is a bug to fix.

File: massbalancemachine/dataloader/DataLoader.py (hunk 4 above)
Comment (JoachimPiret) [reply massbalancemachine/dataloader/DataLoader.py]: When I do the sampling to create an training set and a test set based on "C_region", I'm in fact sampling N times. Since the 12 regions are not homogeneous (some containes more SMB points than other), the sampling is performed N times and then the sampling with the number of points in test/train sets the closest to ts/1-ts is selected.
Each time I do the sampling; train_indices and test_indices are assigned to the dataloader object, ecrasing the previous values. So after the N sampling, the assigned indices corresponds to the last sampling.
I use this function to assign the indices to dataloader corresponding to the selected sampling.

 `  dataloader = mbm.dataloader.DataLoader(cfg, data=data)
  N = 10
  df = pd.DataFrame(index = [f'sampling_{k}' for k in range(1,N+1)],columns = ['percent_train','train_region','train_indices','test_indices'])
  ts = 0.3 ; tf = 'group-c_region'# 'group-rgi' or 'group-c_region'# or 'group-meas-id'
  for i in range(N):`
    
      train_itr, test_itr = dataloader.set_train_test_split(test_size=ts,type_fold = tf,randomness = True) #"group-meas-id"; "group-rgi"; "group-c_region"
      train_indices, test_indices = list(train_itr), list(test_itr)
      df_X_train = data.iloc[train_indices]
      y_train = df_X_train['POINT_BALANCE'].values
      df_X_test = data.iloc[test_indices]
      y_test = df_X_test['POINT_BALANCE'].values

  `train_indices = chosen_df.train_indices.item()
  test_indices = chosen_df.test_indices.item()
  df_X_train = data.iloc[train_indices]
  y_train = df_X_train['POINT_BALANCE'].values
  df_X_test = data.iloc[test_indices]
  y_test = df_X_test['POINT_BALANCE'].values

dataloader.assign_train_test_indices(train_indices, test_indices, ts)`

File: massbalancemachine/data_processing/Dataset.py (hunk 2 above)
Comment (JoachimPiret) [reply massbalancemachine/data_processing/Dataset.py]: yes indeed the np.unique was already applied when creting selff.uniqueID. Thanks !

Comment (JoachimPiret) [reply massbalancemachine/data_processing/Dataset.py]: I checked it again, it returns exactly the same thing as the previous funtion. I removed the redundant np.unique

File: massbalancemachine/dataloader/DataLoader.py [hunk 5]
```diff
@@ -239,14 +264,19 @@ def _prepare_data_for_cv(
         y = train_data["POINT_BALANCE"]
         glacier_ids = train_data["RGIId"].values
         stake_meas_id = train_data["ID"].values  # unique value per stake measurement
-        return X, y, glacier_ids, stake_meas_id
+        try:
+            regions = train_data["C_REGION"].values
+        except:
+            regions = type(np.array([]))
```
Comment (albangossard) [inline massbalancemachine/dataloader/DataLoader.py]: More Pythonic than a try/except which can hide hidden bugs
```suggestion
        regions = train_data["C_REGION"].values if "C_REGION" in train_data.columns else np.array([])
```

File: massbalancemachine/dataloader/DataLoader.py (hunk 4 above)
Comment (albangossard) [reply massbalancemachine/dataloader/DataLoader.py]: Ok I understand better thanks!
Your function `assign_train_test_indices` allows you to set the indices, so it does not have to do anything with the sampling being performed several times. Hence I suggest that you move the sentences which mention this to a note since this can be confusing for the user.

File: massbalancemachine/dataloader/DataLoader.py [hunk 6]
```diff
@@ -108,9 +122,20 @@ def set_train_test_split(
...
-
         return iter(self.train_indices), iter(self.test_indices)
 
+    def assign_train_test_indices(self, train_indices, test_indices, test_size):
+        """
+        Dividing train and test ensemble based on subregion require to make the sampling N times and then choose the
+        train-test division closest to the 70-30 repartition. At each iteration the Dataloader object is redifined as well as
+        self.train_indices and self.test_indices meaning that the information in the Dataloader object are those of the last iterations
+        and not those of the  train-test division chosen after comparing to the 70-30 repartition.
+        This function aims to correct this by reassigning the indices of the chosen sampling.
```
Comment (albangossard) [inline massbalancemachine/dataloader/DataLoader.py]:
```suggestion
        Assign `train_indices`, `test_indices` as well as `test_size` attributes of the object.
        
        Note:
        This can be useful when you divide the train and test ensembles based on subregion since this requires to make the sampling N times and then choose the
        train-test division closest to the 70-30 repartition. At each iteration the Dataloader object is redifined as well as
        self.train_indices and self.test_indices meaning that the information in the Dataloader object are those of the last iterations
        and not those of the  train-test division chosen after comparing to the 70-30 repartition.
        This function aims to correct this by reassigning the indices of the chosen sampling.
```

File: massbalancemachine/data_processing/Dataset.py (hunk 2 above)
Comment (albangossard) [reply massbalancemachine/data_processing/Dataset.py]: You are right thank you!

**Review by albangossard — COMMENTED** [review]
A few minor things to fix, otherwise this looks good to me :)

File: massbalancemachine/dataloader/DataLoader.py [hunk 7]
```diff
@@ -108,9 +122,23 @@ def set_train_test_split(
...
-
         return iter(self.train_indices), iter(self.test_indices)
 
+    def assign_train_test_indices(self, train_indices, test_indices, test_size):
+        """
+        Assign `train_indices`, `test_indices` as well as `test_size` attributes of the object.
+
+        Note:
+        This can be useful when you divide the train and test ensembles based on subregion since this requires to make the sampling N times and then choose the
+        train-test division closest to the 70-30 repartition. At each iteration the Dataloader object is redifined as well as
```
Comment (albangossard) [inline massbalancemachine/dataloader/DataLoader.py]: Just to make it clear since this ratio comes out of the blue
```suggestion
        train-test division closest to, for example, the 70-30 repartition. At each iteration the Dataloader object is redifined as well as
```

File: massbalancemachine/plots/input_plot.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: To be removed, this is not necessary here.

File: massbalancemachine/plots/input_plot.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    ax_xlim : tuple or None, optional
+        x-axis limits as (min, max).
+    ax_ylim : tuple or None, optional
+        y-axis limits as (min, max).
+    test_data : pandas.DataFrame
+        DataFrame containing at least 'target' (observed values) and 'pred' (predicted values) columns for the test data.
+    vvmin, vvmax : float, optional
+        Lower and upper limit of the colormap
+    **kwargs
+        Additional keyword arguments passed to plt.scatter.
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: This argument doesn't exist

File: massbalancemachine/plots/input_plot.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    nbin=100,
+    xlabel="SMB (m/y)",
+    ylabel="Upper surface elevation (m)",
+    ax_xlim=None,
+    ax_ylim=None,
+    test_data=None,
+    vvmin=None,
+    vvmax=None,
+    train_kwargs=None,
+    test_kwargs=None,
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: These two are not used, you can remove them

File: massbalancemachine/plots/input_plot.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    x = grouped_ids["POINT_BALANCE"]
+    y = grouped_ids["POINT_ELEVATION"]
+
+    counts, xedges, yedges = np.histogram2d(x, y, bins=100)
+
+    # Assign each point its bin count
+    ix = np.searchsorted(xedges, x) - 1
+    iy = np.searchsorted(yedges, y) - 1
+    z = counts[ix, iy]
+    if vvmin == None:
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: This is dangerous, better use
```suggestion
    if vvmin is None:
```

File: massbalancemachine/plots/input_plot.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+@author: joachim piret
+"""
+
+import matplotlib.pyplot as plt
+
+# import seaborn as sns
+import numpy as np
+
+
+def scatterplot_training_data(
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: A more explicit name would be great, especially since we may implement similar "scatterplot_training_data" functions.
```suggestion
def scatterplot_mb_elevation(
```

File: massbalancemachine/plots/input_plot.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        ax.set_xlim(ax_xlim)
+    if ax_ylim is not None:
+        ax.set_ylim(ax_ylim)
+
+    plt.tight_layout()
+
+    return fig  # To log figure during training
+
+
+def histogram_training_data(
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: Same thing, the name should be more explicit
```suggestion
def histogram_mb(
```

File: massbalancemachine/plots/input_plot.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    axs=None,
+    title="",
+    xlabel="SMB (m/y)",
+    ylabel="Count",
+    nbins=50,
+    ax_xlim=None,
+    ax_ylim=None,
+    test_data=None,
+    train_kwargs=None,
+    test_kwargs=None,
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: Not used

File: massbalancemachine/plots/input_plot.py [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        Label for the y-axis.
+    nbins : int, optional
+        Number of bins in the histogram.
+    ax_xlim : tuple or None, optional
+        x-axis limits as (min, max).
+    ax_ylim : tuple or None, optional
+        y-axis limits as (min, max).
+    test_data : pandas.DataFrame
+        DataFrame containing at least 'target' (observed values) and 'pred' (predicted values) columns for the test data.
+    **kwargs
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py]: Not used either

File: massbalancemachine/plots/input_plot.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+    Returns
+    -------
+    fig : matplotlib.figure.Figure or None
+        Figure object if a new one was created, otherwise None.
+    """
+
+    if axs is None:
+        if test_data is None:
+            fig, axs = plt.subplots(1, 1, figsize=(10, 5))
```
Comment (albangossard) [inline massbalancemachine/plots/input_plot.py:157]: Great thanks!

File: massbalancemachine/plots/perf_plots.py [hunk 17]
```diff
@@ -112,6 +112,92 @@ def predVSTruth(
     return fig  # To log figure during training
 
 
+def show_training_data(
```
Comment (albangossard) [inline massbalancemachine/plots/perf_plots.py]: As discussed today, this function should be removed.

**Review by albangossard — COMMENTED** [review]
Thanks @JoachimPiret for keeping up that PR!

File: massbalancemachine/plots/input_plot.py (hunk 15 above)
Comment (JoachimPiret) [reply massbalancemachine/plots/input_plot.py]: I removed **kwargs and the related sentence but kept test_data and y-axis... since used in the function

**Review by albangossard — APPROVED** [review]
Congrats @JoachimPiret for your first MBM contribution! 👏
